modules/transport-netty4/src/internalClusterTest/java/org/opensearch/http/netty4/Netty4Http2IT.java

```diff
@@ -44,19 +48,44 @@ public void testThatNettyHttpServerSupportsHttp2() throws Exception {
                 assertThat(responses, hasSize(5));
 
                 Collection<String> opaqueIds = Netty4HttpClient.returnOpaqueIds(responses);
-                assertOpaqueIdsInAnyOrder(opaqueIds);
+                assertOpaqueIdsInAnyOrder(5, opaqueIds);
             } finally {
                 responses.forEach(ReferenceCounted::release);
             }
         }
     }
 
-    private void assertOpaqueIdsInAnyOrder(Collection<String> opaqueIds) {
+    public void testThatNettyHttpServerSupportsHttp2PostUpgrades() throws Exception {
+        final List<Tuple<String, CharSequence>> requests = List.of(Tuple.tuple("/_search", "{\"query\":{ \"match_all\":{}}}"));
+
+        HttpServerTransport httpServerTransport = internalCluster().getInstance(HttpServerTransport.class);
+        TransportAddress[] boundAddresses = httpServerTransport.boundAddress().boundAddresses();
+        TransportAddress transportAddress = randomFrom(boundAddresses);
+
+        try (Netty4HttpClient nettyHttpClient = Netty4HttpClient.http2()) {
+            Collection<FullHttpResponse> responses = nettyHttpClient.post(transportAddress.address(), requests);
+            try {
+                assertThat(responses, hasSize(1));
+
+                for (FullHttpResponse response : responses) {
+                    assertThat(response.getStatus(), equalTo(HttpResponseStatus.OK));
+                }
+
+                Collection<String> opaqueIds = Netty4HttpClient.returnOpaqueIds(responses);
+                String msg = String.format(Locale.ROOT, "Expected opaque id [0], got [%s]", opaqueIds);
+                assertOpaqueIdsInAnyOrder(1, opaqueIds);
+            } finally {
+                responses.forEach(ReferenceCounted::release);
+            }
+        }
+    }
+
+    private void assertOpaqueIdsInAnyOrder(int expected, Collection<String> opaqueIds) {
         // check if opaque ids are present in any order, since for HTTP/2 we use streaming (no head of line blocking)
         // and responses may come back at any order
         int i = 0;
         String msg = String.format(Locale.ROOT, "Expected list of opaque ids to be in any order, got [%s]", opaqueIds);
```

Comment:
Not a big deal, but I think you can remove this because `assertThat` will generate a good error message by default